tools/Google.Cloud.Tools.Analyzers/InternalOptionalParametersRequired.cs

```diff
@@ -0,0 +1,110 @@
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterOperationAction(AnalyzeInvocation, OperationKind.InvocationExpression);
+        }
+
+        private static void AnalyzeInvocation(OperationAnalysisContext context)
+        {
+            // TODO: Move some of this to a helper and share with a code fix provider to auto-insert the missing argument.
+
+            // Only perform the check on calls to externally visible methods which are defined within the same assembly.
+            var invocation = (IInvocationExpression)context.Operation;
+            if (invocation.Syntax is InvocationExpressionSyntax invocationExpression &&
+                invocation.TargetMethod?.IsExternallyVisible() == true &&
+                context.Compilation.Assembly == invocation.TargetMethod.ContainingAssembly)
+            {
+                var semanticModel = context.Compilation.GetSemanticModel(context.Operation.Syntax.SyntaxTree);
+                var variablesInScope =
+                    (from symbol in semanticModel.LookupSymbols(context.Operation.Syntax.SpanStart)
+                     where symbol.Kind == SymbolKind.Local || symbol.Kind == SymbolKind.Parameter
+                     select symbol).ToList();
+
+                foreach (var arg in invocation.ArgumentsInParameterOrder)
+                {
+                    if (arg.ArgumentKind != ArgumentKind.DefaultValue || arg.Parameter == null)
+                    {
+                        continue;
+                    }
+
+                    var parameterType = arg.Parameter.Type;
+                    var convertibleVariables = variablesInScope.Where(
+                        symbol => context.Compilation.ClassifyConversion(symbol.GetVariableType(), parameterType).IsImplicit).ToList();
+                    if (convertibleVariables.Count == 0)
+                    {
+                        continue;
+                    }
+
+                    var nameMatch = convertibleVariables.FirstOrDefault(
+                        symbol => symbol.Name.Equals(arg.Parameter.Name, StringComparison.OrdinalIgnoreCase));
+
+                    // If the parameter's type is String or any primitive type, only show the diagnostic when there is
+                    // some variable with the same (case-insentive) name as the argument. Strings are too commmon, for
+                    // example, to assume that any string in scope should be supplied for a default parameter.
+                    if (parameterType.SpecialType == SpecialType.System_String || parameterType.IsPrimitive())
+                    {
+                        if (nameMatch == null)
```

Comment:
I usually try to avoid changing logical operators in `if` conditions, but this one isn't too bad so I probably would have used a single `if` here. I think I intended to have more logic in here at first but then realized I could just fall through to below and things would work out. I just didn't clean it up after changing course.